src/cljs/s3_beam/client.cljs

```diff
@@ -17,13 +21,43 @@
                (.setParameterValue "file-name" fname)
                (.setParameterValue "mime-type" fmime))))
 
-(defn sign-file [server-url file ch]
-  (let [fmap    (file->map file)
-        edn-ize #(reader/read-string (.getResponseText (.-target %)))]
-    (xhr/send (signing-url server-url (:name fmap) (:type fmap))
-           (fn [res]
-             (put! ch {:f file :signature (edn-ize res)})
-             (close! ch)))))
+(defn sign-file
+  "Takes a `server-url`, either an input map (with keys :file and optionally :identifier and :key),
+  as `input-map-or-file` and a channel `ch`. Also takes a fn `edn-ize` which is for interpreting
+  the server response text into edn.
+  Takes a `key-fn` function that takes one argument (the filename) and generates a object-key to
+  use to create the file on S3.
+  Alternatively, if a :key key exists in the input-map, use that in preference to the key-fn as an object-key.
+  Sends the request to the server-url to be signed."
+  [server-url edn-ize key-fn input-map-or-file ch]
+  (let [xhr      (XhrIo.)
+        {:keys [file identifier] :as input-map}
+                 (if (map? input-map-or-file)
+                   input-map-or-file
+                   {:file input-map-or-file})
+        fmap     (file->map file)
+        fmap     (cond
+                   (:key input-map) (assoc fmap :name (:key input-map))
+                   key-fn           (update-in fmap [:name] key-fn)
+                   :else            fmap)]
+    (events/listen xhr goog.net.EventType.SUCCESS
+      (fn [_]
+        (put! ch {:f file
+                  :signature (edn-ize (.getResponseText xhr))
+                  :identifier identifier})
+        (close! ch)))
+    (events/listen xhr goog.net.EventType.ERROR
+      (fn [_]
+        (let [error-code      (.getLastErrorCode xhr)
+              error-message   (.getDebugMessage goog.net.ErrorCode error-code)
+              http-error-code (.getStatus xhr)]
+          (put! ch {:identifier      identifier
+                    :error-code      error-code
+                    :error-message   (str "While trying to sign file: "
+                                           error-message)
+                    :http-error-code http-error-code})
+          (close! ch))))
+    (. xhr (send (signing-url server-url (:name fmap) (:type fmap)) "GET" nil nil))))
```

Comment:
1. Move the generation ove the signing url into the let binding. 
2. The last three arguments to `send are optional so no need to pass `nil`
3. use `(.send xhr ...)` or `(. xhr send ...)` (the parens around `send` are not necessary I believe)